packages/core/src/store/query/EventQueryImpl.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  NodeId,
+  NodeEventTypes,
+  EditorStore,
+  EventQuery,
+} from '../../interfaces';
+
+export class EventQueryImpl implements EventQuery {
+  private event: Set<NodeId>;
+
+  constructor(private readonly store: EditorStore, type: NodeEventTypes) {
+    this.event = store.getState().events[type];
+  }
+
+  contains(id: NodeId) {
+    return this.event.has(id);
+  }
+
+  getState() {
+    return this.event;
+  }
+
+  getNodes() {
+    const nodes = Array.from(this.event);
+    return nodes.map((id) => this.store.query.node(id));
+  }
+
+  getNodeAtIndex(index: number) {
+    const nodes = this.getNodes();
+    return nodes[index];
+  }
+
+  getFirst() {
+    return this.getNodeAtIndex(0);
+  }
+
+  getLast() {
+    return this.getNodeAtIndex(this.event.size - 1);
```

Comment:
Would that not result in `-1` for a list of size 0? Maybe add a check as safe guard?

Reply:
Good catch